### PR TITLE
Suggestion for solution #476.

### PR DESCRIPTION
**Pull request by klakegg**
Reads files inside data/ and makes data available in .Site.Data. May need some refactoring.

### CODE REVIEW

**Comment by bjornerik** [comment]
Hei og takk for sist!

No har ein også #748 -- men løysinga di er både enkel og fleksibel mht. format. Det blir kanskje heller eit spørsmål om "globale data" eller data som høyrer ei side til

Kan du fjerne den utkommenterte seksjonen og gjere ein "push -f" (overskrive innsjekkinga di).

To others: Sorry about the nynorsk language.

**Comment by klakegg** [comment]
Verden er liten. ;)

Første innsjekk er overskrevet.

And to the english version...
I thing a "global data" solution is wanted in #476, and it's the most appealing solution to me. A "site data" solution could be to create a special directory inside the data directory.

File: hugolib/site.go [hunk 1]
```diff
@@ -332,9 +373,10 @@ func (s *Site) initialize() (err error) {
 	}
 
 	staticDir := helpers.AbsPathify(viper.GetString("StaticDir") + "/")
+	dataDir := helpers.AbsPathify(viper.GetString("StaticDir") + "/")
```
Comment (bjornerik) [inline hugolib/site.go]: -> DataDir

**Comment by bjornerik** [comment]
This should work and is simple enough, @spf13 ?

It should support sub-folders, but that's simple to add.

**Comment by spf13** [comment]
Looks really good. It's pretty much exactly like how I would do it (except in needs tests ;) ) . 

About sub directories, where would they appear?    .Site.Data[subdir][filename] seems like the obvious place to me.

**Comment by bjornerik** [comment]
` .Site.Data[subdir][filename]` is clever -- then it's easy to iterate over all data files for a given subdir, but I suggest using path (i.e. Unix style slashes) and not filepath for the subdir key.

@klakegg do you have any time to spare to take a stab at adding support for subfolders? There are examples of using the `filepath.Walk` some other places in the Hugo code base. Note the ERROR logging on symbolic links -- this might be good to copy over, as this seems to confuse users.

Also some tests would be cool!

**Comment by klakegg** [comment]
I think using Filesystem is better because of less duplication (symlinks) and included caching.

Files in data/ are imported before data in subfolders allowing overriding using subfolders.

**Comment by bjornerik** [comment]
Filesystem vs Walk, I don't care.

But overriding in sub-folders will get confusing pretty fast.

* I think people would expect to be able to create more than one level of subdirs and address the data items in a way that matches that directory structure
* Ideally something like this: `/data/sub1/sub2/mydata.json` ==> `{{ sub1.sub2.mydata.mykey }}`

Any thoughts @spf13 ?

**Comment by klakegg** [comment]
If /data/sub1/sub2.json sets key 'mydata.mykey' to 'foo', and you have a file /data/sub1/sub2/mydata.json setting 'mykey' to 'bar'. What to you expect .Site.Data.sub1.sub2.mydata to return?

Btw, now also with test. ;)

**Comment by bjornerik** [comment]
In that case I would expect /sub2/mydata.json's value, but that isn't override as a feature -- I would even consider returning an error for these duplicates so the user can do some renaming.

**Comment by bjornerik** [comment]
Oh, and the failing test is most likely due to Go maps' random iteration order.

**Comment by klakegg** [comment]
The implementation is now changed to support any amount of subfolders. The test is still unstable (not changed) and will be updated if this is the solution we want.

**Comment by bjornerik** [comment]
There is some spring code cleaning to be done here, but:

* It's simple enough after some limited tests, It does what I would expect.
* I might tone down the duplicate check to just log an ERROR and not STOP -- but that's details

So, @spf13 - to sum up (with my special $ syntax):

`/data/my.json$mkey:1234 => {{ .Data.my.mkey }} ==> 1234`
`/data/sub1/my.json$mkey:1234 => {{ .Data.sub1.my.mkey }} ==> 1234`
`/data/sub1.json$mkey:1234 => {{ .Data.sub1.my.mkey }} ==> 1234 ==> boom ERROR`

It wold need documentation and tests, but from a bird's perspective, it looks correct, right?

**Comment by spf13** [comment]
@bjornerik I don't follow the special $ syntax. Does that just mean the files content after it?

I think we are heading in a very good direction. 

I think duplication should just log an ERROR.

**Comment by bjornerik** [comment]
the $ syntax is just there to separate the file (my.json) with the key inside the file (mkey). It's just for this discussion and not part of any API.

**Comment by spf13** [comment]
Ok. I think this looks really good. We should fix the tests, merge it, and add documentation.

**Comment by bjornerik** [comment]
@klakegg could you fix/add some tests and doc?

**Comment by bjornerik** [comment]
@klakegg bump. It isn't that busy in the government :-) If this is going into the 0.13 release it needs docs/tests -- and I'm too lazy/busy doing it myself.

**Comment by bep** [comment]
OK, I need this one, so I'll finish it.

**Comment by bep** [comment]
Replaced by #885

**Pull request closed without merging**